File: Autonomous Mobile Robots Course/Chapter15/Lesson2/Chapter15_Lesson2.m
% Chapter15_Lesson2.m
% Stanley Controller for Ground Vehicles (kinematic bicycle model, educational)
%
% Run:
%   Chapter15_Lesson2
%
% This script:
%   1) builds a smooth reference path (polyline)
%   2) tracks it using Stanley steering control
%   3) plots trajectory and tracking errors

function Chapter15_Lesson2
    % Reference path
    n = 400;
    xr = linspace(0, 50, n);
    yr = 2.5*sin(0.18*xr) + 1.0*sin(0.04*xr);

    % Parameters
    p.L = 2.7;
    p.k = 1.4;
    p.v0 = 0.5;
    p.maxSteer = deg2rad(32);

    dt = 0.02;  T = 25.0;
    steps = floor(T/dt);

    % State: [x;y;psi], constant speed v
    x = -2.0; y = 3.5; psi = deg2rad(-10);
    v = 6.0;

    xs = zeros(steps,1);
    ys = zeros(steps,1);
    eys = zeros(steps,1);
    epsis = zeros(steps,1);
    deltas = zeros(steps,1);

    for k = 1:steps
        [delta, e_y, e_psi] = stanley_control(x, y, psi, v, xr, yr, p);
        % Bicycle step
        x = x + v*cos(psi)*dt;
        y = y + v*sin(psi)*dt;
        psi = wrap_angle(psi + v/p.L * tan(delta) * dt);

        xs(k) = x; ys(k) = y;
        eys(k) = e_y; epsis(k) = e_psi;
        deltas(k) = delta;
    end

    % Plots
    figure; plot(xr, yr, 'LineWidth', 1.5); hold on;
    plot(xs, ys, 'LineWidth', 1.5);
    axis equal; grid on;
    legend('reference path','vehicle trajectory');
    title('Stanley Controller: path tracking');

    t = linspace(0, T, steps);
    figure; plot(t, eys, 'LineWidth', 1.5); hold on;
    plot(t, rad2deg(epsis), 'LineWidth', 1.5);
    grid on; legend('e_y [m]','e_\psi [deg]');
    title('Tracking errors');

    figure; plot(t, rad2deg(deltas), 'LineWidth', 1.5);
    grid on; legend('\delta [deg]');
    title('Steering command');
end

function [delta, e_y, e_psi] = stanley_control(x, y, psi, v, xr, yr, p)
    % Find nearest path index (simple vertex scan)
    d2 = (xr - x).^2 + (yr - y).^2;
    [~, i] = min(d2);

    % Estimate reference heading using neighbors
    i0 = max(i-1, 1);
    i2 = min(i+1, numel(xr));
    psi_ref = atan2(yr(i2)-yr(i0), xr(i2)-xr(i0));

    % Convention: e_psi = psi_ref - psi
    e_psi = wrap_angle(psi_ref - psi);

    % Signed cross-track via 2D cross product
    tx = cos(psi_ref); ty = sin(psi_ref);
    vx = x - xr(i);   vy = y - yr(i);
    cross = tx*vy - ty*vx;
    e_y = sign_safe(cross) * hypot(vx, vy);

    delta = e_psi + atan2(p.k * e_y, v + p.v0);
    delta = min(max(delta, -p.maxSteer), p.maxSteer);
end

function s = sign_safe(a)
    if a >= 0, s = 1; else, s = -1; end
end

function a = wrap_angle(a)
    a = mod(a + pi, 2*pi) - pi;
end

File: Tutorials/Codes/Autonomous_Mobile_Robots_Course/Chapter15/Lesson2/Chapter15_Lesson2.py
"""
Chapter15_Lesson2.py
Stanley Controller for Ground Vehicles (kinematic bicycle model)
- Path: polyline (x_ref, y_ref)
- Control: delta = delta_ff + e_psi + atan2(k * e_y, v + v0)
Author: Abolfazl Mohammadijoo (educational example)
"""

from __future__ import annotations
import math
from dataclasses import dataclass
from typing import Tuple

import numpy as np
import matplotlib.pyplot as plt


def wrap_angle(a: float) -> float:
    """Wrap angle to (-pi, pi]."""
    return (a + math.pi) % (2.0 * math.pi) - math.pi


@dataclass
class State:
    x: float
    y: float
    psi: float  # yaw (rad)
    v: float    # forward speed (m/s)


@dataclass
class StanleyParams:
    L: float = 2.7                 # wheelbase (m)
    k: float = 1.2                 # cross-track gain
    v0: float = 0.5                # softening speed (m/s) to avoid divide-by-zero
    max_steer: float = math.radians(30.0)  # steering saturation (rad)


def _nearest_point_on_polyline(px: float, py: float, x_ref: np.ndarray, y_ref: np.ndarray) -> Tuple[int, float, float, float]:
    """
    Return:
      idx: segment index i such that nearest point lies on segment i->i+1 (or endpoint)
      nx, ny: nearest point coordinates
      t: projection parameter in [0,1] on that segment
    """
    x1 = x_ref[:-1]
    y1 = y_ref[:-1]
    x2 = x_ref[1:]
    y2 = y_ref[1:]
    dx = x2 - x1
    dy = y2 - y1
    denom = dx*dx + dy*dy
    denom = np.where(denom < 1e-12, 1e-12, denom)

    t = ((px - x1)*dx + (py - y1)*dy) / denom
    t = np.clip(t, 0.0, 1.0)

    projx = x1 + t*dx
    projy = y1 + t*dy
    d2 = (px - projx)**2 + (py - projy)**2

    i = int(np.argmin(d2))
    return i, float(projx[i]), float(projy[i]), float(t[i])


def _segment_heading(i: int, x_ref: np.ndarray, y_ref: np.ndarray) -> float:
    dx = float(x_ref[i+1] - x_ref[i])
    dy = float(y_ref[i+1] - y_ref[i])
    return math.atan2(dy, dx)


def _curvature_at_index(i: int, x_ref: np.ndarray, y_ref: np.ndarray) -> float:
    """
    Discrete curvature estimate around point i using finite differences on indices.
    Robust enough for educational demos; for production, compute curvature vs arc-length.
    """
    n = len(x_ref)
    i0 = max(0, i-1)
    i1 = i
    i2 = min(n-1, i+1)
    x0, y0 = x_ref[i0], y_ref[i0]
    x1, y1 = x_ref[i1], y_ref[i1]
    x2, y2 = x_ref[i2], y_ref[i2]

    # Approx derivatives w.r.t index (not arc length)
    x_p = (x2 - x0) * 0.5
    y_p = (y2 - y0) * 0.5
    x_pp = x2 - 2.0*x1 + x0
    y_pp = y2 - 2.0*y1 + y0

    denom = (x_p*x_p + y_p*y_p)**1.5
    if denom < 1e-9:
        return 0.0
    kappa = (x_p*y_pp - y_p*x_pp) / denom
    return float(kappa)


def stanley_control(state: State, x_ref: np.ndarray, y_ref: np.ndarray, params: StanleyParams) -> Tuple[float, dict]:
    """
    Compute Stanley steering command.
    Convention:
      e_psi = psi_ref - psi   (desired minus actual)
      e_y   = signed cross-track error (positive if vehicle is left of path tangent)
    """
    i, nx, ny, _ = _nearest_point_on_polyline(state.x, state.y, x_ref, y_ref)
    psi_ref = _segment_heading(i, x_ref, y_ref)
    e_psi = wrap_angle(psi_ref - state.psi)

    # Signed lateral error via 2D cross product between tangent and (vehicle - nearest)
    tx = math.cos(psi_ref)
    ty = math.sin(psi_ref)
    vx = state.x - nx
    vy = state.y - ny
    cross = tx*vy - ty*vx  # >0 => vehicle left of tangent
    e_y = math.copysign(math.hypot(vx, vy), cross)

    # Optional feedforward from reference curvature
    kappa_ref = _curvature_at_index(i, x_ref, y_ref)
    delta_ff = math.atan(params.L * kappa_ref)

    # Stanley feedback term
    delta_fb = e_psi + math.atan2(params.k * e_y, state.v + params.v0)

    delta = delta_ff + delta_fb
    delta = max(-params.max_steer, min(params.max_steer, delta))

    info = {
        "i": i,
        "nx": nx,
        "ny": ny,
        "psi_ref": psi_ref,
        "e_psi": e_psi,
        "e_y": e_y,
        "kappa_ref": kappa_ref,
        "delta_ff": delta_ff,
        "delta_fb": delta_fb,
    }
    return delta, info


def step_bicycle(state: State, delta: float, dt: float, L: float) -> State:
    x = state.x + state.v * math.cos(state.psi) * dt
    y = state.y + state.v * math.sin(state.psi) * dt
    psi = wrap_angle(state.psi + state.v / L * math.tan(delta) * dt)
    return State(x=x, y=y, psi=psi, v=state.v)


def make_s_path(n: int = 400) -> Tuple[np.ndarray, np.ndarray]:
    # A simple smooth polyline path
    x = np.linspace(0.0, 50.0, n)
    y = 2.5 * np.sin(0.18 * x) + 1.0 * np.sin(0.04 * x)
    return x, y


def main() -> None:
    x_ref, y_ref = make_s_path()

    params = StanleyParams(L=2.7, k=1.4, v0=0.5, max_steer=math.radians(32))
    dt = 0.02
    T = 25.0
    steps = int(T / dt)

    state = State(x=-2.0, y=3.5, psi=math.radians(-10), v=6.0)

    xs, ys, psis, deltas = [], [], [], []
    eys, epsis = [], []

    for _ in range(steps):
        delta, info = stanley_control(state, x_ref, y_ref, params)
        state = step_bicycle(state, delta, dt, params.L)

        xs.append(state.x)
        ys.append(state.y)
        psis.append(state.psi)
        deltas.append(delta)
        eys.append(info["e_y"])
        epsis.append(info["e_psi"])

    # Plots
    plt.figure(figsize=(9, 5))
    plt.plot(x_ref, y_ref, label="reference path")
    plt.plot(xs, ys, label="vehicle trajectory")
    plt.axis("equal")
    plt.grid(True)
    plt.legend()
    plt.title("Stanley Controller: path tracking (kinematic bicycle)")
    plt.xlabel("x [m]")
    plt.ylabel("y [m]")

    t = np.linspace(0.0, T, steps)
    plt.figure(figsize=(9, 4))
    plt.plot(t, eys, label="e_y [m]")
    plt.plot(t, np.degrees(epsis), label="e_psi [deg]")
    plt.grid(True)
    plt.legend()
    plt.title("Tracking errors")
    plt.xlabel("time [s]")

    plt.figure(figsize=(9, 4))
    plt.plot(t, np.degrees(deltas), label="delta [deg]")
    plt.grid(True)
    plt.legend()
    plt.title("Steering command")
    plt.xlabel("time [s]")

    plt.show()


if __name__ == "__main__":
    main()
